src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -71,99 +83,280 @@ public long BytesWritten
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> can still be re-used after disposing but all its state has been reset.
+        /// </remarks>
+        public async ValueTask DisposeAsync()
+        {
+            await FlushAsync().ConfigureAwait(false);
+            ResetHelper();
+        }
+#endif
+
+        /// <summary>
+        /// Asynchronously commits the JSON text written so far which makes it visible to the output destination.
+        /// </summary>
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it asynchronously, while monitoring cancellation requests.
+        /// </remarks>
+        public async Task FlushAsync(CancellationToken cancellationToken = default)
+        {
+            FlushHelper();
+
+            if (_stream != null)
+            {
+                Debug.Assert(_arrayBufferWriter != null);
+                Debug.Assert(BytesPending == _arrayBufferWriter.WrittenCount);
+                if (BytesPending != 0)
+                {
+#if BUILDING_INBOX_LIBRARY
+                    await _stream.WriteAsync(_arrayBufferWriter.WrittenMemory, cancellationToken).ConfigureAwait(false);
+#else
+                    Debug.Assert(_arrayBufferWriter.WrittenMemory.Length == _arrayBufferWriter.WrittenCount);
+                    await _stream.WriteAsync(_arrayBufferWriter.WrittenMemory.ToArray(), 0, _arrayBufferWriter.WrittenCount, cancellationToken).ConfigureAwait(false);
+#endif
+                    _arrayBufferWriter.Clear();
+                }
+                await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            _memory = default;
+            BytesCommitted += BytesPending;
+            BytesPending = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void FlushHelperStream()
+        {
+            Debug.Assert(_arrayBufferWriter != null);
+            Debug.Assert(BytesPending == _arrayBufferWriter.WrittenCount);
+            if (BytesPending != 0)
+            {
+#if BUILDING_INBOX_LIBRARY
+                _stream.Write(_arrayBufferWriter.WrittenSpan);
+#else
+                Debug.Assert(_arrayBufferWriter.WrittenSpan.Length == _arrayBufferWriter.WrittenCount);
+                _stream.Write(_arrayBufferWriter.WrittenSpan.ToArray(), 0, _arrayBufferWriter.WrittenCount);
+#endif
+                _arrayBufferWriter.Clear();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void FlushHelper()
+        {
+            _output.Advance(BytesPending);
```

Comment:
For the stream mode, should I use the concrete type `_arrayBufferWriter.Advance()` rather than going through an interface call like this?

Reply:
Why do we have this helper at all?  It's a single line in a method marked AggressiveInlining.

Reply:
Artifact of merging two concepts/terminology. With stream/IBW mode, we have "advance IBW" and we have "write/flush to stream", which effectively have the same semantics as far as the JsonWriter is concerned so I normalized it. This just helped me make sense of what operations I am doing while reading the source (normalizes to call Flush... for both cases).

This goes back to my original question of should I call `_output.Advance` in the stream case or call `_arrayBufferWriter.Advance`?

Do you think I should remove it regardless?

Reply:
I suggest you remove it and then at the call sites instead do whatever is the more efficient call.